homeassistant/components/fyta/sensor.py

```diff
@@ -0,0 +1,180 @@
+    ),
+    FytaSensorEntityDescription(
+        key="salinity_status",
+        translation_key="salinity_status",
+        device_class=SensorDeviceClass.ENUM,
+        options=PLANT_STATUS_LIST,
+        value_fn=lambda value: PLANT_STATUS[value],
+    ),
+    FytaSensorEntityDescription(
+        key="temperature",
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+        device_class=SensorDeviceClass.TEMPERATURE,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    FytaSensorEntityDescription(
+        key="light",
+        translation_key="light",
+        native_unit_of_measurement="mol/d",
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    FytaSensorEntityDescription(
+        key="moisture",
+        native_unit_of_measurement=PERCENTAGE,
+        device_class=SensorDeviceClass.MOISTURE,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    FytaSensorEntityDescription(
+        key="salinity",
+        translation_key="salinity",
+        native_unit_of_measurement="mS/cm",
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    FytaSensorEntityDescription(
+        key="ph",
+        device_class=SensorDeviceClass.PH,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    FytaSensorEntityDescription(
+        key="battery_level",
+        native_unit_of_measurement=PERCENTAGE,
+        device_class=SensorDeviceClass.BATTERY,
+        state_class=SensorStateClass.MEASUREMENT,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+    FytaSensorEntityDescription(
+        key="plant_number",
+        translation_key="plant_number",
+        state_class=SensorStateClass.MEASUREMENT,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the FYTA binary sensors."""
+    coordinator: FytaCoordinator = hass.data[DOMAIN][entry.entry_id]
+
+    plant_entities: list[CoordinatorEntity] = []
+    plant_entities = [
+        FytaCoordinatorSensor(coordinator, entry, sensor)
+        for sensor in SENSORS
+        if sensor.key in coordinator.data or sensor.key == "plant_number"
+    ]
+
+    plants = coordinator.plant_list
+    plant_entities.extend(
+        [
+            FytaPlantSensor(coordinator, entry, sensor, plant_id)
+            for plant_id in plants
+            for sensor in SENSORS
+            if sensor.key in coordinator.data[plant_id]
+        ]
+    )
+
+    async_add_entities(plant_entities)
+
+
+class FytaCoordinatorSensor(FytaCoordinatorEntity, SensorEntity):
+    """Represents a Fyta sensor."""
+
+    entity_description: FytaSensorEntityDescription
+
+    @property
+    def native_value(self) -> str | int | float | datetime:
+        """Return the state for this sensor."""
+
+        if "plant_number" in self.entity_description.key:
+            val = len(self.coordinator.plant_list)
+        else:
```

Comment:
This can also be solved via value_fn

Reply:
Usually we just pass the coordinator(.data) to the value_fn. This way the logic to actually fetch the data is in the sensor entity descriptions. (Since otherwise you would have a big list of `if key == "a": return data.a elif key == "b": return data.b`). This way we keep the class as generic as possible and keep the logic of 1 sensor in 1 place.

currently I am not sure what sensors should spawn at what device exactly and which can appear at both places. Maybe it makes sense to have a PlantSensorEntityDescription, with a function that takes the plant as parameter. The native_value property would only contain `self.entity_description.value_fn(self.plant)`. The value_fn would then be `value_fn=lambda plant: plant.something`.